Automattic/newspack-ads
# feat: Ads providers and Broadstreet integration

Add support for multiple providers alongside GAM for integration with the placements system. Once #316 is implemented, it'll be the core method for displaying ads through Blocks and Widgets as well.

This PR also adds [Broadstreet](https://broadstreetads.com/) as a core provider, which is automatically enabled if [its plugin](https://wordpress.org/plugins/broadstreet/) is installed and active. It uses methods from the [plugin's `Broadstreet_Utility` class](https://github.com/broadstreetads/broadstreet-wp/blob/master/Broadstreet/Utility.php).

<img width="817" alt="image" src="https://user-images.githubusercontent.com/820752/156440823-68a10e05-50b8-4ad0-9ec6-99c5e3240228.png">

Some notes:
 - I have moved the GAM API class to the `providers/gam` directory for better organization
 - The existence of this new system calls for a refactor of the `blocks` and `model` classes, which mixes core methods with GAM-specific helpers. I plan to address this refactor in a different PR since it won't bring new features and its best to be reviewed separately.

Closes #288 

### How to test this PR

1. While on the master branch, make sure you have GAM connected and placements configured with valid ad units
2. Switch to this branch and https://github.com/Automattic/newspack-plugin/pull/1521
3. Confirm the previously configured placements continues to render as expected
4. Edit a placement and confirm the new select control with "Google Ad Manager" (If you don't have Broadstreet installed, it should only have GAM as an option)
5. Install Broadstreet and ping me for demo account access
6. Select a Broadstreet zone and confirm it renders with both AMP Plus on and off and AMP deactivated
7. Written tests should also pass
